File: sorting/dutch_national_flag_problem.js
/*
    Dutch National Flag Problem

    Given an array containing only 0s, 1s, and 2s, sort the array in a single traversal.

    Sample Input: [0, 2, 1, 2, 0]
    Sample Output: [0, 0, 1, 2, 2]

    Approach:
    - We can use three pointers, low, mid, and high, to divide the array into three regions:
      1. 0s region: elements before the low pointer (excluding low) are 0s
      2. 1s region: elements between the low and mid pointers (excluding mid) are 1s
      3. 2s region: elements after the high pointer (excluding high) are 2s
    - Initialize low and mid pointers to the start of the array (0 index) and high pointer to the end of the array (array.length - 1).
    - Iterate while the mid pointer is less than or equal to the high pointer:
      - If the current element at mid is 0, swap it with the element at low and increment both low and mid pointers.
      - If the current element at mid is 1, it is already in the correct region, so we just increment the mid pointer.
      - If the current element at mid is 2, swap it with the element at high and decrement the high pointer.
    - Repeat the above steps until the mid pointer crosses the high pointer.
    - At the end, the array will be sorted in place.

    Time Complexity: O(n), where n is the length of the array.
    Space Complexity: O(1), no additional space is used.

    Further Reading: https://en.wikipedia.org/wiki/Dutch_national_flag_problem
*/

function dutchNationalFlagProblem(arr) {
  let low = 0;
  let mid = 0;
  let high = arr.length - 1;

  while (mid <= high) {
    if (arr[mid] === 0) {
      // Swap current element at mid with element at low
      [arr[mid], arr[low]] = [arr[low], arr[mid]];
      // Increment both low and mid pointers
      low++;
      mid++;
    } else if (arr[mid] === 1) {
      // Move to the next element in the 1s region
      mid++;
    } else {
      // Swap current element at mid with element at high
      [arr[mid], arr[high]] = [arr[high], arr[mid]];
      // Decrement the high pointer
      high--;
    }
  }

  return arr;
}

// Test the function
const input = [0, 2, 1, 2, 0];
const output = dutchNationalFlagProblem(input);
console.log(output);

File: sorting/dutch_national_flag_problem.java
/*
 * Dutch National Flag Problem
 *
 * Given an array of 0's, 1's, and 2's, sort the array in linear time, without using extra space.
 *
 * This is known as the Dutch National Flag problem.
 *
 * The idea here is to divide the array into three regions such that the starting region contains all the 0's,
 * the next region contains all the 1's, and the final region contains all the 2's.
 *
 * The algorithm works by maintaining three pointers:
 *
 *   * `low`: The index of the first 0 in the array.
 *   * `mid`: The index of the current element being processed.
 *   * `high`: The index of the last 2 in the array.
 *
 * The algorithm then iterates through the array, and does the following:
 *
 *   * If the current element is a 0, swap it with the element at `low` and increment both `low` and `mid` by 1.
 *   * If the current element is a 1, increment `mid` by 1.
 *   * If the current element is a 2, swap it with the element at `high` and decrement `high` by 1.
 *
 * Once the algorithm has finished iterating, the array will be sorted.
 */

public class DutchNationalFlag {

    public static void sort(int[] array) {
        int low = 0;
        int mid = 0;
        int high = array.length - 1;

        while (mid <= high) {
            switch (array[mid]) {
                case 0:
                    swap(array, low, mid);
                    low++;
                    mid++;
                    break;
                case 1:
                    mid++;
                    break;
                case 2:
                    swap(array, mid, high);
                    high--;
                    break;
            }
        }
    }

    private static void swap(int[] array, int i, int j) {
        int temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }

    public static void main(String[] args) {
        int[] array = {1, 0, 2, 1, 0, 2, 0, 1};

        sort(array);

        for (int i = 0; i < array.length; i++) {
            System.out.print(array[i] + " ");
        }

        System.out.println();
    }
}
